text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1029 @@
+        (for l @in self { l }, for r @in r { r })
+    }
+}
+
+// Reverse a tuple's members
+trait Rev {
+    type Value;
+    fn rev(self) -> Self::Value;
+}
+impl Rev for () { 
+    type Value = ();
+    fn rev(self) -> Self::Value { () }
+}
+impl<Head, (..Tail)> Rev for (Head, ..Tail) 
+where
+    (..Tail): Rev,
+    <(..Tail) as Rev>::Value: Merge<(Head,)> {
+    type Value = <<(..Tail) as Rev>::Value as Merge<(Head,)>>::Output;
+
+    fn rev(self) -> Self::Value { 
+        let (h, t @ ..) = self; 
+        let rev_t = <(..Tail) as Rev>::rev(t);
+        rev_t.merge((h,)) 
+    }
+}
+
+// Utility to create a tuple of a single type, for instance: ToT<usize> = (..T) -> (..usize)
+// We could have a syntaxic sugar to do this (future RFC?)
+trait ToT<T> {
+    type Value = T;
+}
+impl<T, A> ToT<T> for A { }
+
+
+// Example usage
+// Reverse the tuple and provide a tuple with the hashes of the tuple members
+fn reverse_tuple_and_hash<(..T), (..RevT)>(value: (..T)) -> (<(..T) as Rev>::Value, (..<T as ToT<usize>>::Value)),
+where
+    (..T): Rev<Value = (..RevT)>,
+    ..(T: Hash), {
+
+    let rev_t = <(..T) as Rev>::rev(value);
+
+    // Here we use the identifiers of the reversed variadic tuple and variadic tuple type in the iteration
+    let hashes = (for ref rev_t <RevT> in rev_t <RevT> { 
+        let mut s = DefaultHasher::new();
+        <RevT as Hash>::hash(rev_t, &mut s);
+        s.finish()
+    });
+
+    (
+        (for rev_t @in rev_t { rev_t }), 
+        hashes,
+    )
+}
+```
+
+## Declaring and using multiple variadic tuple type with same arity
+
+In C++ multiple parameter packs can be expanded in a single expansion form as long as the packs have the same number of items, but there is no constraint concerning the declaration.
+
+For Rust, using the syntax `(..(T1, T2, ..., Tn))` embeds the constraint that the variadic tuple types `T1`, `T2`, ..., `Tn` have the same arity. This is more consistent than not grouping the declaration (ie: `(..T1), (..T2), ..., (..Tn))`) because the signature using the declaration contains all the information required.
+
+We don't need to look at the implementation or body code to know the required constraint about the variadic tuple type arities.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+C++11 sets a decent precedent with its variadic templates, which can be used to define type-safe variadic functions, among other things. This RFC is comparable to the design of _type parameter packs_ (variadic tuple type) and _function parameter pack_ (variadic tuple).
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+## Dynamic libraries tuple implementation assumption
+
+When using dynamic libraries, client libraries may relies that the host contains code up to a specific tuple arity. So we need to have a 
+way to enforce the compiler to generate all the implementation up to a specific tuple arity. (12 will keep backward comptibility with current `std` impl)
```

Comment:
Huh?  This needs a code sample for clarification.

Reply:
Consider a host application A which load a dynamic library B.

Let's consider A as:
```rust
fn main() {
      // Load dynamic library B
      load_dylib();
      // Execute an entry point that was defined by the inventory crate
      execute_entry_point();
}
```

And dylib B:
```rust
// Custom macro to register the entry point fn
register_entry_point!(entry_point);

fn entry_point() {
    let a = (1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1); // 13-arity tuple with all uint
    let b = (1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1); // Second operand
    println!("a = b {}", a == b); // We need PartialEq impl for 13-arity tuple.
}
```

While compiling A, there will be included in the host implementation of `PartialEq` for tuples with arity up to 12. (If we force to have at least 12-arity tuple implementations).

B don't include the implementation of `PartialEq` for tuple as it is a dylib, but requires its implementation for 13-arity tuples.

So, when A load B, B will be using an implementation that is not loaded.
This will be an issue.

Reply:
Hm, I'm still not sure that I see it.  I only see two possibilities for the above snippet, depending on whether PartialEq support for 13-ary tuples exists in rust at the time of dylib B's compilation:
* Either the impl *can* be generated at the time of dylib B's compilation (in which case it will be included in the object files for dylib B, and will work regardless of how the dylib is used)
* ...or it *cannot* be generated, and dylib B cannot be compiled.

Basically, I can't see anything like this ever being an issue is because a dylib doesn't even *include* impls for e.g. arbitrary tuples of size 3.  It couldn't, because there exist an infinite number of such types.  Code for generic functions and implementations in Rust is generated at precisely the point when they are known to be needed, which is the point whenever Rust knows the concrete types of all of the generic type arguments to a function call.

Reply:
>Either the type does impl PartialEq at the time of dylib B's compilation (in which case the machine code for the corresponding definition of PartialEq::eq will be included in the object files for dylib B, and will work regardless of how the dylib is used)

I think this where my assumption was incorrect, My understanding was that the machine code was not included in dylib B as it is a trait defined in core and to avoid duplicate machine code when loading multiple dylibs.

So, this is not an issue as the code will be embedded in dylibs, I can remove this section